evm/src/memory/segments.rs

```diff
@@ -0,0 +1,22 @@
+/// Contains EVM bytecode.
+pub const CODE: usize = 0;
+
+pub const STACK: usize = 1;
+
+/// Main memory, owned by the contract code.
+pub const MAIN_MEM: usize = 2;
+
+/// Memory owned by the kernel.
+pub const KERNEL_MEM: usize = 3;
+
+/// Data passed to the current context by its caller.
+pub const CALLDATA: usize = 4;
+
+/// Data returned to the current context by its latest callee.
+pub const RETURNDATA: usize = 5;
+
+/// A segment which contains a few fixed-size metadata fields, such as the caller's context, or the
+/// size of `CALLDATA` and `RETURNDATA`.
+pub const METADATA: usize = 6;
+
+pub const NUM_SEGMENTS: usize = 7;
```

Comment:
Hm would we list them out separately in the assert? What do you think of using an enum instead, like

```
pub enum Segment {
    Code = 0,
    Stack = 1,
    ...
}
```

Seems a bit safer since it wouldn't rely on a separate list. It wouldn't check `NUM_SEGMENTS` though; if we wanted that we could use one of those crates with a procedural macro for number of variants.